### PR TITLE
Post processing

### LINKED ISSUE #78: New Rule - Functions access private VNET resources
Add post processing to access app settings
Write rule for function apps integrated with a VNET so that its setting are 
* properties.WEBSITE_DNS_SERVER: 168.63.129.16
* properties.WEBSITE_VNET_ROUTE_ALL: 1

### PR DESCRIPTION
**Pull request by Joshua-Phelps**
closes #78 

* adds post processing though RequestEvaluation
* Adds function app rule
* Adds ARM template for function app environment in separate Resource Group 
    * one correctly configured 
    * one misconfigured
* removes nested deployment in original ARM template  
* Adds tests for function app
* fixes global before and after hooks
* changed key vault id to use env password instead

### CODE REVIEW

File: docs/built-in-rules.md [hunk 1]
```diff
@@ -35,4 +35,19 @@ Finds Event Hubs with a network rule set having one or more IP and virtual netwo
...
 
 See [Event Hubs - Allow IP Addresses - ARM Template](https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-ip-filtering#use-resource-manager-template) for adding IP firewall rules and setting the defaultAction.   
-See [Event Hubs - Allow VNETS - ARM Template](https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-service-endpoints#use-resource-manager-template) for adding VNET firewall rules and setting the defaultAction.  
\ No newline at end of file
+See [Event Hubs - Allow VNETS - ARM Template](https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-service-endpoints#use-resource-manager-template) for adding VNET firewall rules and setting the defaultAction.  
+
+## Function App VNET Integration Misconfiguration
+name: "function-app-vnet-integration-misconfiguration"
+### Description 
+Finds Function Apps integrated with a virtual network but the app settings for WEBSITE_VNET_ROUTE_ALL is not 1 and WEBSITE_DNS_SERVER is not 168.63.129.16.
```
Comment (noelbundick) [inline docs/built-in-rules.md:43]:
```suggestion
Finds Function Apps integrated with a virtual network but the app settings for `WEBSITE_VNET_ROUTE_ALL` is not `1` and `WEBSITE_DNS_SERVER` is not `168.63.129.16`. 
```

File: MY_TEMPLATE.json [hunk 2]
```diff
@@ -0,0 +1,1424 @@
+{
```
Comment (noelbundick) [inline MY_TEMPLATE.json:1]: remove

File: rules.json [hunk 3]
```diff
@@ -19,5 +19,41 @@
...
+      "query": "type == `Microsoft.EventHub/namespaces/networkRuleSets` && properties.defaultAction == `Allow` && (length(properties.ipRules) > `0` || length(properties.virtualNetworkRules) > `0`)"
+    }
+  },
+  {
+    "name": "function-app-vnet-integration-misconfiguration",
+    "description": "",
+    "type": "ARM",
+    "recommendation": "https://github.com/noelbundick/config-analyzer/blob/main/docs/built-in-rules.md#function-app-vnet-integration-misconfiguration",
+    "evaluation": {
+      "query": "type == `Microsoft.Web/sites`",
```
Comment (Joshua-Phelps) [inline rules.json:47]: add kind

Comment (noelbundick) [inline rules.json:47]: add kind: functionapp

File: src/rules/armTemplate.ts [hunk 4]
```diff
@@ -75,14 +106,32 @@ export class ARMTemplateRule implements BaseRule<ARMTarget> {
...
       resources: ['*'],
       options: 'SkipAllParameterization',
     });
+    return {
+      subscriptionId,
+      groupName,
+      credential,
+      client,
+      template: template._response.parsedBody.template,
+      type: 'ARM' as RuleType.ARM,
```
Comment (noelbundick) [inline src/rules/armTemplate.ts:119]: cleanup?

File: src/rules/armTemplate.ts [hunk 5]
```diff
@@ -38,12 +41,40 @@ type AndEvaluation = BaseEvaluation & {
...
+}
+
+function mapAsync<T1, T2>(
+  array: T1[],
+  callback: (value: T1, index: number, array: T1[]) => Promise<T2>
+): Promise<T2[]> {
+  return Promise.all(array.map(callback));
+}
+
+async function filterAsync<T>(
```
Comment (noelbundick) [inline src/rules/armTemplate.ts:68]: nit: comments?

File: src/rules/armTemplate.ts [hunk 6]
```diff
@@ -75,14 +106,32 @@ export class ARMTemplateRule implements BaseRule<ARMTarget> {
...
+      template: template._response.parsedBody.template,
+      type: 'ARM' as RuleType.ARM,
+    };
   }
 
-  execute(target: ARMTarget) {
-    const results = this.evaluate(this.evaluation, target.template);
+  async execute(target: ARMTarget) {
+    let results = this.evaluate(this.evaluation, target.template);
+    const evaluation = this.evaluation;
```
Comment (noelbundick) [inline src/rules/armTemplate.ts:125]: nit: duplicate name/object?

File: src/rules/armTemplate.ts [hunk 7]
```diff
@@ -75,14 +106,32 @@ export class ARMTemplateRule implements BaseRule<ARMTarget> {
...
+
+    // If we found resources from the first evaluations, filter those down to the ones that meet all the criteria for the request evaluation
+    if (results.length > 0 && isRequestEvaluation(evaluation)) {
```
Comment (noelbundick) [inline src/rules/armTemplate.ts:128]: use only type guard?

File: src/rules/armTemplate.ts [hunk 8]
```diff
@@ -38,12 +41,40 @@ type AndEvaluation = BaseEvaluation & {
...
+  callback: (value: T1, index: number, array: T1[]) => Promise<T2>
+): Promise<T2[]> {
+  return Promise.all(array.map(callback));
+}
+
+async function filterAsync<T>(
+  array: T[],
+  callback: (value: T, index: number, array: T[]) => Promise<boolean>
+): Promise<T[]> {
+  const filterMap = await mapAsync(array, callback);
```
Comment (noelbundick) [inline src/rules/armTemplate.ts:72]: check for edge cases / undefined?

File: src/rules/armTemplate.ts [hunk 9]
```diff
@@ -113,26 +162,48 @@ export class ARMTemplateRule implements BaseRule<ARMTarget> {
...
+  async sendRequest(
+    target: ARMTarget,
+    resource: ARMResource,
+    evaluation: RequestEvaluation
+  ) {
+    const token = await target.credential.getToken(
+      'https://graph.microsoft.com/.default'
+    );
+    const options = {
+      url: `https://management.azure.com/subscriptions/${target.subscriptionId}/resourceGroups/${target.groupName}/providers/${resource.type}/${resource.name}/${evaluation.request.operation}?api-version=${resource.apiVersion}`,
```
Comment (noelbundick) [inline src/rules/armTemplate.ts:196]: ensure we have test coverage?

File: src/rules/armTemplate.ts [hunk 10]
```diff
@@ -113,26 +162,48 @@ export class ARMTemplateRule implements BaseRule<ARMTarget> {
...
+      method: 'POST' as HttpMethods,
```
Comment (noelbundick) [inline src/rules/armTemplate.ts:197]: use enum value?

File: src/rules/armTemplate.ts [hunk 11]
```diff
@@ -113,26 +162,48 @@ export class ARMTemplateRule implements BaseRule<ARMTarget> {
...
+      headers: {
+        Authorization: `Bearer ${token?.token}`,
+        'Content-Type': 'application/json',
+        Host: 'management.azure.com',
```
Comment (noelbundick) [inline src/rules/armTemplate.ts:201]: nit: is this extra?

File: test/azure/rules/armTemplate.spec.ts [hunk 12]
```diff
@@ -47,4 +55,43 @@ describe('ARM Template Rule', function () {
...
+  it('tests the function app misconfiguration with the RequestEvaluation', async () => {
+    const target = await ARMTemplateRule.getTarget(
+      subscriptionId,
+      resourceGroup2,
+      credential
+    );
+    const rule = new ARMTemplateRule({
+      name: 'function-app-vnet-integration-misconfiguration',
+      description: '',
+      type: 'ARM' as RuleType.ARM,
```
Comment (noelbundick) [inline test/azure/rules/armTemplate.spec.ts:67]: fix?

**Review by noelbundick — APPROVED** [review]
lgtm w/ comments

**Pull request merged**